### ISSUE TITLE
Make RiemannSurfacePath __repr__ and __str__ more meaningful

### ISSUE THREAD

**Issue by cswiercz** [cswiercz/abelfunctions#69]
There are two ways to print Python objects using the class methods `__repr__` and `__str__`:

* The goal of `__repr__` is to be *unambiguous*.
* The goal of `__str__` is to be *readable*.

That is, class instances should be distinguishable from their `__repr__` output. Optionally, the `__str__` output can be used to make the output more readable.

Applied to Riemann surface paths, the output could be something like this:

* `__repr__` - output the underlying Riemann surface and a list of tuples representing the path.
    * For cycles, a path can be represented by a list of discriminant points of the curve and the number of rotations around each discriminant point. (A la Maple.) For example,

    ```python
    # this cycle goes around the discriminant point at 
    # x=0 twice in the positive direction and the point
    # at x=I three times in the negative direction
    gamma = X.a_cycles()[0]
    print gamma
    ```
    ```
    Cycle on the Riemann surface defined by f(x,y) = (the curve): [(0.0, 2), (1.0j, -3)].
    ```
    Another option would be to simply output
    ```
    [(0.0, 2), (1.0j, -3)]
    ```
    but this would not be unique with respect to choice of Riemann surface; not as much of a problem considering that people will often work with one surface at a time. One thing missing from this representation is the starting `y`-value.

    * For paths (non-cycles) this would be messy. Sure, the part of the path used for sheet changing can be represented using the cycle `__repr__` output. However, the remainder of the path would have to use a different notation. Alternatively, the entire path can be represented by a list of lines, semicircles, and rays: the primitive geometric objects used to construct paths on Riemann surfaces. For example,

    ```python
    # this path consists of a straight line between x=0
    # and x=(1/2 + i), a semicircle around x=1+i of
    # radius 1/2 starting at theta=-pi and travelling
    # positive pi radians (positively oriented semicircle;
    # and finally a straight line from x=3/2+i to x = 2+2i
    gamma = X.a_cycles()[0]
    print gamma
    ```
    ```
    Path on the Riemann surface defined by f(x,y) = (the curve): [(0,0.5 + 1.0j), (0.5, 1.0 + 1.0j, -3.1415..., 3.1415...), (1.5 + 1.0j, 2.0 + 2.0j)]
    ```
    Similarly, we can simply output:
    ```
    [(0,0.5 + 1.0j), (0.5, 1.0 + 1.0j, -3.1415..., 3.1415...), (1.5 + 1.0j, 2.0 + 2.0j)]
    ```
    but unless people carefully read the documentation this is meaningless. It also suffers the issue of not knowing the starting `y`-value.

* `__str__` - we can use `__repr__` as above. Or, we can output something that hides all the specific details. I don't think something too different from above is necessary. Another option is to make `__str__` return the strings discussed above and `__repr__` can output the usual class instance with memory location stuff.

A key issue here is that there does not exist a `RiemannSurfaceCycle` class separate from the general `RiemannSurfacePath` class. A cycle is simply an `RSPath` that has the same start and end point.

**Comment by cswiercz** [comment]
> 1. I think you shouldn't worry (yet) about how to present output for non cycles: for paths from P1 to P2. That's a different animal. 

I'm having trouble figuring out the best way to go around doing this mostly because of my last comment above:

*"A key issue here is that there does not exist a `RiemannSurfaceCycle` class separate from the general `RiemannSurfacePath` class. A cycle is simply an `RSPath` that has the same start and end point."*

In Maple a cycle really is just a list of sheets, branch points, and rotation numbers. There isn't a separate class / type representing them. Rather, there are functions dedicated to interpreting lists of this form as cycles and working with the list data accordingly. In abelfunctions, all paths are of type `RiemannSurfacePath`. There isn't a separate "Cycle" type: they too are `RiemannSurfacePath` objects but they happen to have the same starting and ending point.

This may sound like a non-issue but from a design point of view it doesn't seem to make sense to use two different representations, depending on the state of the object, for the same type of object. I need to read up on this design choice.
 
> 2: a small modification of one of your proposed outputs would be to give 
>
> (P1, B1, P2, B2, P3, ...)
>
> where 
> P1 is the basepoint with sheet (x,y1)
> B1 is the branchpoint with circulation number (xb,n)
> P2 is the basepoint with sheet (x,y2)
> ...
>
> It's a bit overdetermined, but it makes it easy for people to check stuff. 

It turns out that I already have a function that returns the kind information you wanted, but it's not a "top-level" function to `RiemannSurface`. It's used internally to construct these `RiemannSurfacePath` objects.

```python
g = X.genus()
a_homology, b_homology, c_homology, linear_combinations = X.PF.YPF.homology()

print 'a-cycles:'
for a_cycle in a_homology:
    print '\ta-cycle:', a_cycle
    
print '\nc-cycles:'
for c_cycle in c_homology:
    print '\tc-cycle:', c_cycle
    
print '\nLinear combinations of c-cycles:'
print linear_combinations
```
```
a-cycles:
	a-cycle: [((-1+0j), 1), (-1j, -1)]
	a-cycle: [(1j, 1), (oo, -1)]
	a-cycle: [((-1+0j), -1), ((1+0j), 1)]
	a-cycle: [(1j, -1), (-1j, -1), (oo, -1)]

c-cycles:
	c-cycle: [((-1+0j), 1), (-1j, -1)]
	c-cycle: [(1j, 1), (-1j, -1)]
	c-cycle: [((-1+0j), -1), (-1j, -2)]
	c-cycle: [(1j, -1), (-1j, -2)]
	c-cycle: [(oo, 1), (-1j, -1)]
	c-cycle: [((1+0j), -1), (-1j, -2)]
	c-cycle: [(0j, 1), (-1j, -1)]
	c-cycle: [((1+0j), 1), (-1j, -1)]
	c-cycle: [(oo, -1), (-1j, -2)]
	c-cycle: [(0j, -1), (-1j, -2)]

Linear combinations of c-cycles:
[[ 1  0  0  0  0  0  0  0  0  0]
 [ 0  1  0  0 -1  0  0  0  0  0]
 [ 0  0  1  0  0 -1  0  0  0  0]
 [ 0  0  0  1 -1  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  1  0]
 [-1  0  0  0  1  0  0  0 -1  0]
 [-1 -1  0  0  1  0  0  1 -1  0]
 [ 0 -1 -1  0  1  0  0  1  0  0]]
```

The "cycles", in this case, really are lists of the form `[..., (branch point, number of rotations), ...]`. These are used internally to help construct the `RiemannSurfacePath` objects and have a representation similar to Maple's `homology()` command.

I can easily make this accessible from `RiemannSurface`. That is, something like

```python
X.homology()
```

can output the same above information, but this might be confusing with the existing `X.a_cycles()`, etc.

**Comment by cswiercz** [comment]
After some discussions with @danshapero it would make sense, first of all, to have `__repr__` be uniform across cycles and non-cycles.

It would also be a *very bad idea* to create a separate class for cycles because doing so would likely run us into the [Circle-Ellipse Problem](http://en.wikipedia.org/wiki/Circle-ellipse_problem). Though, `RiemannSurfacePath` objects can probably be made immutable which is one type of "solution" to the problem.

A possibility for `__repr__` would be to return a list of the `__repr__`s of the primitive constituents of a path:
* for `RiemannSurfacePathLine` output `'Line(start, end)'`,
* for `RiemannSurfacePathArc` output `'Arc(radius, center, theta0, dtheta)'`,
* for `RiemannSurfacePathRay` output `'Ray(start)'` (all rays currently point away from the origin).
Though, some mention of the underlying `RiemannSurface` is probably necessary.

For non-cycles, the output of `__str__` can be something like:

    'Path from (starting place) to (ending place) on the Riemann surface defined by the algebraic curve ...'

It's on the long side, so maybe the "Riemann surface defined by..." part can be omitted.

**Comment by cswiercz** [comment]
Here's a demonstration of what's implemented in ca83b90:

In [1]:
```python
%matplotlib inline

from sympy.abc import x,y
from abelfunctions import (RiemannSurface, RiemannTheta, Jacobian,
                           RiemannConstantVector, puiseux, AbelMap)

f = x**2*y**3 - x**4 + 1
X = RiemannSurface(f,x,y)
print X
```
```
Riemann surface defined by f(x,y) = -x**4 + x**2*y**3 + 1
```
In [2]:
```python
# printing behavior of a cycle
cycle = X.a_cycles()[0]
print cycle
```
```
Cycle on the Riemann surface defined by f(x,y) = -x**4 + x**2*y**3 + 1
```
In [3]:
```python
print cycle.__repr__()
```
```
RiemannSurfacePath(Line((-2+0j),(-1.2+0j)),Arc((0.2+0j),(-1+0j),(-3.14159265359+0j),(3.14159265359+0j)),Arc((0.2+0j),(-1+0j),0j,(3.14159265359+0j)),Line((-1.2+0j),(-2+0j)),Line((-2+0j),(-0.2-1j)),Arc((0.2+0j),-1j,(-3.14159265359+0j),(-3.14159265359+0j)),Arc((0.2+0j),-1j,0j,(-3.14159265359+0j)),Line((-0.2-1j),(-2+0j))) on the Riemann surface defined by f(x,y) = -x**4 + x**2*y**3 + 1
```
In [4]:
```python
# printing behavior of a path to a regular place
places = X(2)
P = places[0]
print P
```
```
(2, RootOf(4*_y**3 - 15, 0))
```
In [5]:
```python
gamma = X.path(P)
print gamma
```
```
Path from ((-2+0j), (-0.7768081264884648+1.3454711428104118j)) to ((2+0j), (1.5536162529769295+0j)) on the Riemann surface defined by f(x,y) = -x**4 + x**2*y**3 + 1
```
In [6]:
```python
print gamma.__repr__()
```
```
RiemannSurfacePath(Line((-2+0j),(-0.2-1j)),Arc((0.2+0j),-1j,(-3.14159265359+0j),(-3.14159265359+0j)),Arc((0.2+0j),-1j,0j,(-3.14159265359+0j)),Line((-0.2-1j),(-2+0j)),Line((-2+0j),(-1.2+0j)),Arc((0.2+0j),(-1+0j),(3.14159265359+0j),(-3.14159265359+0j)),Line((-0.8+0j),(-0.2+0j)),Arc((0.2+0j),0j,(3.14159265359+0j),(-3.14159265359+0j)),Line((0.2+0j),(0.8+0j)),Arc((0.2+0j),(1+0j),(3.14159265359+0j),(-3.14159265359+0j)),Line((1.2+0j),(2+0j))) on the Riemann surface defined by f(x,y) = -x**4 + x**2*y**3 + 1
```
In [7]:
```python
# printing behavior of a path to a discriminant place
places = X(0)
P = places[0]
eta = X.path(P)
print eta
```
```
Path from ((-2+0j), (-0.7768081264884648+1.3454711428104118j)) to (-t**3, -1/t**2 + t**10/3 + t**22/9 + O(t**31)) on the Riemann surface defined by f(x,y) = -x**4 + x**2*y**3 + 1
```
In [8]:
```python
print eta.__repr__()
```
```
RiemannSurfacePath(Line((-2+0j),(-0.2-1j)),Arc((0.2+0j),-1j,(-3.14159265359+0j),(3.14159265359+0j)),Arc((0.2+0j),-1j,0j,(3.14159265359+0j)),Line((-0.2-1j),(-2+0j)),Line((-2+0j),(-1.2+0j)),Arc((0.2+0j),(-1+0j),(3.14159265359+0j),(-3.14159265359+0j)),Line((-0.8+0j),(-0.2+0j)),Line((-0.2+0j),0j)) on the Riemann surface defined by f(x,y) = -x**4 + x**2*y**3 + 1
```

The `__repr__`s are verbose but they're, in a sense, supposed to be that way.

**Comment by cswiercz** [comment]
I'll label this as closed for now. See the commit message in 6000db738583f6f07f493dd842abad20c5a9a3be for some thoughts on why I consider this closed.
